pipeline/dags/download_images.py

```diff
@@ -0,0 +1,255 @@
+"""
+Depth-first parallelized download procedure that creates a DAG for each collection.
+Collections are split into chunks and downloaded into disk images.
+"""
+
+from datetime import timedelta
+from pathlib import Path
+from datetime import date
+import os
+import json
+
+from airflow.operators.empty import EmptyOperator
+from airflow.providers.http.sensors.http import HttpSensor
+from airflow.providers.ssh.hooks.ssh import SSHHook
+from airflow.hooks.base import BaseHook
+from airflow.decorators import task, task_group, dag
+
+from harvester.pmh_interface import PMH_API
+from harvester import utils
+
+from importlib import reload
+
+reload(utils)
+
+from operators.custom_operators import (
+    SaveMetsSFTPOperator,
+    SaveAltosSFTPOperator,
+)
+
+INITIAL_DOWNLOAD = True
+
+BASE_PATH = "/scratch/project_2006633/nlf-harvester/images"
+TMPDIR = "/local_scratch/robot_2006633_puhti/harvester-temp"
+IMAGE_SPLIT_DIR = Path("/home/ubuntu/image_split/")
+BINDING_BASE_PATH = Path("/home/ubuntu/binding_ids_all")
+SSH_CONN_ID = "puhti_conn"
+HTTP_CONN_ID = "nlf_http_conn"
+SET_IDS = ["col-361"]
+
+default_args = {
+    "owner": "Kielipankki",
+    "start_date": "2023-05-22",
+    "retry_delay": timedelta(minutes=5),
+    "retries": 3,
+}
+
+http_conn = BaseHook.get_connection(HTTP_CONN_ID)
+api = PMH_API(url=http_conn.host)
+
+
+def read_bindings(set_id):
+    try:
+        with open(BINDING_BASE_PATH / set_id / f"binding_ids_{date.today()}", "r") as f:
+            bindings = f.read().splitlines()
+        return bindings
+    except FileNotFoundError:
+        print(
+            "No binding file found for today. Make sure to run DAG 'fetch_binding_ids' first."
+        )
+        return []
```

Comment:
This exists solely to prevent Airflow UI from not loading DAGs due to the "error" that the binding fetching function was not run that day. In our real use case we first run the binding fetching DAG and the download DAG only after, but since Airflow executes the code in the DAG definition files every n minutes, this is one option to stop that. The other option would be to schedule the binding fetching DAG to run every day and make sure to keep it unpaused always.

Reply:
I implemented the latter one, and set the start date for checking new bindings to be the last run of the _download DAG_ of the set instead of the ID fetching DAG.